### ISSUE TITLE
Implementation of Unsafe Class

### ISSUE THREAD

**Issue by arakis** [mosa/MOSA-Project#745]
If we provide an Implementation of Unsafe class and ref structs, this will a big boost into the right direction and makes low level system programming much more type safe and easier.

The most important functions, in the order:
```
public static int SizeOf<T>()
public static ref TTo As<TFrom, TTo>(ref TFrom source)
public static void* AsPointer<T>(ref T value)
```
If the above methods are implemented correctly, the following methods should work "automatically", because they depend on the above methods:
```
public static ref T AsRef<T>(void* source)
public static void Write<T>(void* destination, T value)
public static T Read<T>(void* source)
```

It could be necessary to provide the `ByReference<T>` Type.

Please keep in mind that SizeOf<T>() should NOT rely on metadata lookup for each call, it will be horrible slow. As I know, Mosa is duplicating for every generic call the code - so compiler should know the size of T at compile time. And if the methods are correctly implemented as Intrinsics, they could be inlined within the calling method, wich means, that nearly every Call to the above methods is done in very less (maybe only a single) instruction, especially the ref struct methods.

After all the above stuff is working, the following features will work without any additional Compiler support:

- removing unsafe-pointers and replace them via ref structs
- ref structs
- Span<T>
- ReadonlySpan<T>
- Memory<T>
- And much more cool stuff.

Here's a small reference implementation incl. a test scenario at the end of the file:
https://github.com/abanu-org/abanu/blob/fd6a45042722596198763b9b63c6b92484d73dc2/src/Abanu.Lib/BitReader.cs

It would be nice if this feature could get a reasonable priority.

**Comment by tgiphil** [comment]
Partial implementation for SizeOf<T>:

https://github.com/mosa/MOSA-Project/blob/master/Source/Mosa.Compiler.Framework/Stubs/Unsafe.cs#L11

**Comment by tgiphil** [comment]
@charsleysa Is this completed?

**Comment by charsleysa** [comment]
> @charsleysa Is this completed?

Not quite, I think some of the read and write methods are still missing, maybe roughly 80% complete.

**Comment by AnErrupTion** [comment]
This is mostly done, just a few functions (2 I think) and the Unsafe class is fully implemented.
